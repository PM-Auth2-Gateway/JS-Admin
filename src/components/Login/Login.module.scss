@media(max-width: 767px){
  .dropdown{
    right: auto;
  }
}