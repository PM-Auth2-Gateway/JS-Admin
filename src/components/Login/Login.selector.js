import { createSelector, createStructuredSelector } from 'reselect';
const socialsState = (state) => state.socialsList;
const userState = (state) => state.user;

const socialsSelector = createSelector(socialsState, ({ socials }) => socials);

const userSelector = createSelector(userState, ({ user }) => user);

const isUserLoadingSelector = createSelector(
  userState,
  ({ isLoading }) => isLoading
);

const authenticatedSelector = createSelector(
  userState,
  ({ authenticated }) => authenticated
);

export const loginSelector = createStructuredSelector({
  socials: socialsSelector,
  isUserLoading: isUserLoadingSelector,
  authenticated: authenticatedSelector,
  user: userSelector,
});
