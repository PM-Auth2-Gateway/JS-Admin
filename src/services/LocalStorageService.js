export default class LocalStorageService {
  static getToken() {
    return localStorage.getItem('token');
  }

  static setToken(token) {
    localStorage.setItem('token', token);
  }
}
