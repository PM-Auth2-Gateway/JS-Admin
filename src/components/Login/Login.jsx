import React, { useEffect } from 'react';
import { Dropdown, Image } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import { getSocials } from '../../ducks/auth/socials';
import { loginSelector } from './Login.selector';
import {
  login,
  logout,
  setAuthenticated,
  setUser,
} from './../../ducks/auth/user';
import LocalStorageService from './../../services/LocalStorageService';

const LoginButton = () => {
  const dispatch = useDispatch();
  const { socials, authenticated, user } = useSelector(loginSelector);

  useEffect(() => {
    dispatch(getSocials());
    if (LocalStorageService.getToken()) {
      dispatch(setAuthenticated(true));
      dispatch(setUser(LocalStorageService.getUser()));
    }
  }, [dispatch]);

  const handleLogin = (id) => {
    dispatch(login(id));
  };

  const handleLogout = () => {
    dispatch(logout());
  };

  return (
    <Dropdown className='mr-5'>
      {!authenticated && !user ? (
        <>
          <Dropdown.Toggle variant='primary'>Login</Dropdown.Toggle>
          <Dropdown.Menu>
            {socials &&
              socials.map((soc) => (
                <Dropdown.Item key={soc.id} onClick={() => handleLogin(soc.id)}>
                  {soc.name}
                </Dropdown.Item>
              ))}
          </Dropdown.Menu>
        </>
      ) : (
        <>
          <Dropdown.Toggle variant='light'>
            <Image
              src={user.photo}
              alt={user.firstName}
              width='30'
              h='30'
              roundedCircle
              className='mr-1'
            />
            {user.firstName} {user.lastName}
          </Dropdown.Toggle>
          <Dropdown.Menu>
            <Dropdown.Item onClick={handleLogout}>Logout</Dropdown.Item>
          </Dropdown.Menu>
        </>
      )}
    </Dropdown>
  );
};

export default LoginButton;
