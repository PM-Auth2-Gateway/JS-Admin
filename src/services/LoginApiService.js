import { instance } from '../axiosConfig';
import LocalStorageService from './LocalStorageService';

export default class LoginApiService {
  static app_id = 1;
  static session_id;
  static async getSocials() {
    const { data } = await instance.get('Socials/', {
      headers: {
        App_id: this.app_id,
      },
    });
    return data.socials;
  }

  static async getAuthLink(social_id) {
    const { data } = await instance.post(
      'Socials/auth-link',
      { social_id, device: 'browser' },
      {
        headers: {
          App_id: LoginApiService.app_id,
        },
      }
    );
    LoginApiService.session_id = data.state;
    return this.buildUrl(data);
  }

  static async getProfile() {
    const { data } = await instance.post(
      'Admin/tokenAndProfile',
      {
        session_id: LoginApiService.session_id,
      },
      {
        headers: {
          App_id: LoginApiService.app_id,
        },
      }
    );
    LocalStorageService.setToken(data.token);
    return data;
  }

  static buildUrl(authLink) {
    const {
      auth_uri,
      redirect_uri,
      response_type,
      client_id,
      scope,
      state,
    } = authLink;

    return `${auth_uri}?redirect_uri=${redirect_uri}&response_type=${response_type}&client_id=${client_id}&scope=${scope}&state=${state}`;
  }

  static async login(social_id) {
    const link = await LoginApiService.getAuthLink(social_id);
    const modal = window.open(
      link,
      'Auth',
      'width=972,height=660,modal=yes,alwaysRaised=yes'
    );

    const interval = setInterval(() => {
      if (modal.closed) {
        clearInterval(interval);
      }
    }, 1000);
  }
}
