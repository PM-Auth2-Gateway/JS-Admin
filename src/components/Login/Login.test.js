import React from 'react';
import Login from './Login';
import { shallow, mount } from 'enzyme';
import { useSelector, useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';
import LocalStorageService from './../../services/LocalStorageService';
import LoginApiService from '../../services/LoginApiService';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  useHistory: jest.fn(),
}));

jest.mock('./../../services/LocalStorageService');

jest.mock('./../../services/LoginApiService');
describe('Login component', () => {
  const dispatch = jest.fn();
  const historyPush = jest.fn();

  const data = {
    socials: [{ id: 1, name: 'test' }],
    authenticated: false,
    user: null,
  };

  const authData = {
    ...data,
    authenticated: true,
    user: {
      firstName: 'first',
      lastName: 'last',
      photo: 'test.png',
    },
  };

  beforeEach(() => {
    useDispatch.mockReturnValue(dispatch);
    useSelector.mockReturnValue(data);
    useHistory.mockReturnValue({
      push: historyPush,
    });
  });

  it('should render', () => {
    const component = shallow(<Login />);
    expect(component).toMatchSnapshot();
  });

  it('should render user data if user authenticated', () => {
    useSelector.mockReturnValue(authData);

    const component = shallow(<Login />);
    expect(component).toMatchSnapshot();
  });

  it('should redirect to root page on logout', () => {
    useSelector.mockReturnValue(authData);
    const component = shallow(<Login />);
    const logout = component.find('DropdownItem');
    logout.simulate('click');
    expect(historyPush).toBeCalledWith('/');
  });

  it('should call dispatch on first render', () => {
    mount(<Login />);
    expect(dispatch).toBeCalledTimes(1);
  });

  it('should call dispatch 3 time on first render if user and token present in localstorage', () => {
    LocalStorageService.getToken.mockReturnValue('token');
    mount(<Login />);
    expect(dispatch).toBeCalledTimes(2);
  });

  it('should fetch authLink on social click', () => {
    const component = shallow(<Login />);
    const socialBtn = component.find('DropdownItem');
    socialBtn.simulate('click');
    expect(LoginApiService.getAuthLink.mock.calls.length).toBe(1);
  });
});
