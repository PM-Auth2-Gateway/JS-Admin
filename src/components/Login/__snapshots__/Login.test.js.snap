// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component should render 1`] = `
<Dropdown
  className="mr-5"
  navbar={false}
>
  <DropdownToggle
    className="d-flex align-items-center"
    variant="primary"
  >
    Login with
  </DropdownToggle>
  <DropdownMenu
    align="left"
    alignRight={false}
    className="dropdown dropdown-menu-right"
    flip={true}
  >
    <DropdownItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      className="d-flex align-items-center"
      disabled={false}
      key="1"
      onClick={[Function]}
    >
      <Image
        className="mr-3"
        fluid={false}
        height="20"
        rounded={false}
        roundedCircle={false}
        thumbnail={false}
        width="20"
      />
      test
    </DropdownItem>
  </DropdownMenu>
</Dropdown>
`;

exports[`Login component should render user data if user authenticated 1`] = `
<Dropdown
  className="mr-5"
  navbar={false}
>
  <DropdownToggle
    className="d-flex align-items-center"
    variant="light"
  >
    <Image
      alt="first"
      className="mr-1"
      fluid={false}
      h="30"
      referrerPolicy="no-referrer"
      rounded={false}
      roundedCircle={true}
      src="test.png"
      thumbnail={false}
      width="30"
    />
    first
     
    last
  </DropdownToggle>
  <DropdownMenu
    align="left"
    alignRight={false}
    className="dropdown dropdown-menu-right"
    flip={true}
  >
    <DropdownItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      disabled={false}
      to="/applications"
    >
      Applications
    </DropdownItem>
    <DropdownItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      data-test="logout-btn"
      disabled={false}
      onClick={[Function]}
    >
      Logout
    </DropdownItem>
  </DropdownMenu>
</Dropdown>
`;
